File: impls/perl/step0_repl.pl
#!perl
use strict;
use warnings;
use Term::ReadLine;

sub READ {
    return $_[0];
}
sub EVAL {
    return $_[0];
}
sub PRINT {
    return $_[0];
}

sub rep {
    my $line=shift;
    return PRINT EVAL READ $line;
}

my $line;
my $term=Term::ReadLine->new('mal');
my $prompt='user> ';
my $OUT=$term->OUT || \*STDOUT;
while(defined ($line=$term->readline($prompt))) {
    chomp $line;
    $term->addhistory($line) if $line =~ /\S/;
    print $OUT $line, "\n";
}

File: impls/js/step0_repl.js
const readline = require('readline');

function READ(line) {
    return line;
}
function EVAL(line) {
    return line;
}
function PRINT(line) {
    return line;
}
function repl(line) {
    return PRINT(EVAL(READ(line)));
}

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  prompt: 'user> '
});

rl.prompt();

rl.on('line', (line) => {
    console.log(line);
    rl.prompt();
}).on('close', () => {
  process.exit(0);
});
